extensions/avro/src/main/java/com/hazelcast/jet/avro/AvroSerializerHooks.java

```diff
@@ -19,42 +19,157 @@
+                    String schemaJson = in.readString();
+                    Schema schema = jsonToSchema.computeIfAbsent(schemaJson,
+                            json -> new Schema.Parser().parse(json));
+                    byte[] serializedDatum = in.readByteArray();
+                    switch (schema.getType()) {
+                        case RECORD:
+                            return new LazyImmutableRecord(serializedDatum, schema);
+                        case ARRAY:
+                            return new LazyImmutableArray<>(serializedDatum, schema);
+                        case ENUM:  // GenericEnumSymbol
+                        case FIXED: // GenericFixed
+                            return deserialize(new GenericDatumReader<>(schema), serializedDatum);
+                        default:
+                            throw new UnsupportedOperationException("Schema type " + schema.getType() + " is unsupported");
+                    }
+                }
+            };
+        }
 
         @Override
-        public Class<Utf8> getSerializationType() {
-            return Utf8.class;
+        public boolean isOverwritable() {
+            return false;
+        }
+    }
+
+    /**
+     * Replaces {@link Utf8} with {@link String} globally. Alternatively,
+     * the string class can be set on a per-schema basis via <ol>
+     * <li> {@link GenericData#setStringType GenericData.setStringType}{@code (schema,}
+     *      {@link StringType#String String}{@code )} or,
+     * <li> {@code SchemaBuilder.record("name").prop(}{@link GenericData#STRING_PROP STRING_PROP}{@code ,}
+     *      {@link StringType#String String}{@code )}.
+     */
+    public static final class CharSequenceHook implements SerializerHook<CharSequence> {
+        @Override
+        public Class<CharSequence> getSerializationType() {
+            return CharSequence.class;
         }
 
         @Override
         public Serializer createSerializer() {
-            return new StreamSerializer<Utf8>() {
+            return new ByteArraySerializer<CharSequence>() {
                 @Override
                 public int getTypeId() {
                     return SerializerHookConstants.AVRO_UTF8;
                 }
 
                 @Override
-                public void write(ObjectDataOutput out, Utf8 object) throws IOException {
-                    out.writeByteArray(object.getBytes());
+                public byte[] write(CharSequence cs) {
+                    return cs instanceof Utf8 ? ((Utf8) cs).getBytes()
+                            : cs.toString().getBytes(StandardCharsets.UTF_8);
                 }
 
                 @Override
-                public Utf8 read(ObjectDataInput in) throws IOException {
-                    return new Utf8(in.readByteArray());
+                public CharSequence read(byte[] buffer) {
+                    return new String(buffer, StandardCharsets.UTF_8);
```

Comment:
this serialization is not reversible. Deserialization always creates String. Will it not cause problems in code which expects `Utf8`, which:
> Unlike String, instances are mutable.

Reply:
I assumed that `Utf8` is solely used in Avro containers, which have dedicated serializers that protect inner `Utf8` instances from this serializer. However, I admit that despite its plausibility, such an assumption may be dangerous to make, and at this point, I decided to switch to the less invasive solutions given in the class Javadoc.